feat!: Update deps and enable new rules 2025-09-24

amannn opened the pull request:
Resolves https://github.com/tools-aoeur/eslint-config/issues/29#event-19870029027

**New rules:**
  - [`prefer-expect-type-of`](https://github.com/hbc1729/eslint-plugin-vitest/blob/main/docs/rules/prefer-expect-type-of.md) (autofixable)
  - [`hoisted-apis-on-top`](https://github.com/vitest-dev/eslint-plugin-vitest/blob/main/docs/rules/hoisted-apis-on-top.md) (needs manual fixing)
- [`preserve-caught-error`](https://eslint.org/docs/latest/rules/preserve-caught-error) (needs manual fixing)

amannn commented on package.json:
```diff
@@ -39,9 +39,9 @@
     "eslint-plugin-react-hooks": "^5.2.0",
     "eslint-plugin-sort-destructure-keys": "^2.0.0",
-    "eslint-plugin-unicorn": "^60.0.0",
+    "eslint-plugin-unicorn": "^61.0.2",
```
New rules that were added:

- [`prefer-bigint-literals`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-bigint-literals.md)
- [`prefer-classlist-toggle`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-classlist-toggle.md)
- [`require-module-attributes`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/require-module-attributes.md)
- [`no-array-sort`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-sort.md)

I personally don't see too much value in them currently.

The other part is the new `unopinionated` config. Looking at the [included rules](https://github.com/sindresorhus/eslint-plugin-unicorn/tree/main?tab=readme-ov-file#rules), a total of 91 out of 110 rules are enabled here.

Looking at a few rules, these seem quite opinionated to me, see e.g. [`expiring-todo-comments`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/expiring-todo-comments.md). It might take significant time to discuss individual rules, evaluate their impact on our projects and opt-out of rules that we don't want.

For the time being I'd not add this.

jdehaan commented on package.json:
```diff
@@ -28,7 +28,7 @@
     "@eslint/js": "^9.34.0",
     "@typescript-eslint/utils": "^8.41.0",
-    "@vitest/eslint-plugin": "^1.3.4",
+    "@vitest/eslint-plugin": "^1.3.12",
```
Yes, please, enable these. I like rules that have a normalizing character. Regarding the hoisting, it leads to a natural and clearer view on what is executed. Nobody should have interest in misleading code requiring js runtime expertise to know if it's running error-free or not... The other rule also ensures a nicer and clear to read form for the same outcome.

jdehaan reviewed (changes requested):
Thx for reading through the changes and digesting these for us! I added comments for my reasons and in short fully share your judgment and would like to have the helpful rules added.

amannn commented:
@jdehaan Thank you for your review! I've enabled the new rules as discussed and changed the PR title to bump the major version due to new rules being enabled that are not autofixable (related to the [versioning policy](https://github.com/tools-aoeur/eslint-config/?tab=readme-ov-file#versioning)).

I also noticed that there were some TS errors in the repo, so I've fixed them too and enabled checking them on CI.

Can you have another look if this looks good to you? Thanks!

jdehaan commented on tsconfig.self.json:
```diff
@@ -0,0 +1,19 @@
+  "compilerOptions": {
+    "target": "es2022",
+    "module": "ESNext",
+    "moduleResolution": "Bundler",
```
👍🏻
